fairlearn/postprocessing/_threshold_optimizer.py

```diff
@@ -427,6 +260,167 @@ def _threshold_optimization_equalized_odds(sensitive_features, labels, scores, g
+    :type flip: bool
+
+    :param plot: Show ROC/selection-error plot if True
+    :type plot: bool
+
+    :param random_state: set to a constant for reproducibility
+    :type random_state: int, np.RandomState, or None
+    """
+
+    def __init__(self, *, estimator=None, constraints=DEMOGRAPHIC_PARITY,
+                 grid_size=1000, flip=True, plot=False, random_state=None):
+
+        self.constraints = constraints
+        self.grid_size = grid_size
+        self.flip = flip
+        self.plot = plot
+        self.estimator = estimator
+        self.random_state = random_state
+
+    def fit(self, X, y, *, sensitive_features, **kwargs):
+        """Fit the model.
+
+        The fit is based on training features and labels, sensitive features, as
+        well as the fairness-unaware predictor or estimator. If an estimator was
+        passed in the constructor this fit method will call `fit(X, y,
+        **kwargs)` on said estimator.
+
+        :param X: The feature matrix
+        :type X: numpy.ndarray or pandas.DataFrame
+
+        :param y: The label vector
+        :type y: numpy.ndarray, pandas.DataFrame, pandas.Series, or list
+
+        :param sensitive_features: sensitive features to identify groups by,
+            currently allows only a single column
+        :type sensitive_features: currently 1D array as numpy.ndarray, list,
+            pandas.DataFrame, or pandas.Series
+        """
+        if self.constraints not in _SUPPORTED_CONSTRAINTS:
+            raise ValueError(NOT_SUPPORTED_CONSTRAINTS_ERROR_MESSAGE)
+        threshold_optimizer = _SUPPORTED_CONSTRAINTS[self.constraints]
+
+        if not is_classifier(self.estimator):
+            raise ValueError("{} needs a classifier to work on, and {} is not "
+                             "one.".format(self.__class__.__name__,
+                                           self.estimator.__class__.__name__))
+
+        X, y, sensitive_features = self._validate_input_data(
+            X, sensitive_features, y)
+
+        self.estimator_ = self.estimator.fit(X, y, **kwargs)
+
+        scores = _get_soft_predictions(self.estimator_, X)
+        self._post_processed_predictor_by_sensitive_feature = threshold_optimizer(
+            sensitive_features, y, scores, self.grid_size, self.flip, self.plot)
+
+    def predict(self, X, *, sensitive_features):
+        """Predict label for each sample in X while taking into account sensitive features.
+
+        :param X: feature matrix
+        :type X: numpy.ndarray or pandas.DataFrame
+
+        :param sensitive_features: sensitive features to identify groups by, currently allows
+            only a single column
+        :type sensitive_features: currently 1D array as numpy.ndarray, list, pandas.DataFrame,
+            or pandas.Series
+
+        :return: predictions in numpy.ndarray
+        """
+        check_is_fitted(self)
+        random_state = check_random_state(self.random_state)
+
```

Comment:
I would prefer having
`self.random_state_ = check_random_state(self.random_state)` in `fit` and then relying on it in `predict`.
That seems to be a better way to implement a randomized predictor.